### ISSUE TITLE
Letsencrypt wildcard certificate problem

### ISSUE THREAD

**Issue by smitpatelx** [tomav/docker-mailserver#1341]
## Context

Using letsencrypt wildcard certificate which can work on both <example.com> and <*.example.com>. The certificate is valid.

## Expected Behavior

I expect email clients not to throw certificate invalid error while setting up

## Actual Behavior

My mail Client is showing `Certificate not valid` with both **SSL/TLS** and **STARTTLS**.
Error log also describes:
```
 failed: error:14094416:SSL routines:ssl3_read_bytes:sslv3 alert certificate unknown: SSL alert number 46
```

## Possible Fix

I dont have any idea of a possible fix. I am confused on using **example.com** or **mail.example.com** for smtp and imap. I tried both in this case but failed.

## Steps to Reproduce

1. map your letsencrypt to /etc/letsencrytp. Mine includes a wildcard certificate
2. **.env** file should contain host `mail` and domain `example.com`
3. docker-compose up -d mail
4. Make sure you have a directory name mail.<your domain>.com in your **`/etc/letsencrypt/live/`** directory.

## Your Environment

* Amount of RAM available: 4 GB
* Mailserver version used: latest
* Docker version used: Docker version 19.03.5, build 633a0ea838
* Environment settings relevant to the config: 
```
HOSTNAME=mail
DOMAINNAME=<domain>.com
CONTAINER_NAME=mail_server
```
* Any relevant stack traces:
```
## /var/log/mail/mail.log

mail dovecot: imap-login: Disconnected (no auth attempts in 0 secs): user=<>, rip=99.248.220.156, lip=172.28.0.2, TLS handshaking: SSL_accept() failed: error:14094416:SSL routines:ssl3_read_bytes:sslv3 alert certificate unknown: SSL alert number 46, session=<Wi/CBcWZZpxj+Nyc>
```

**Comment by smitpatelx** [comment]
## Solution

- Make sure you are not overriding domain name and host in **env-mailserver**
- If you do so it smtp wont work.
- Their must be a **MX** record for `mail.<domain>.com` pointing to `mail.<domain>.com`
- An **A** Record pointing to your host ipv4 in name of `mail.<domain>.com`.
- Make sure your `/etc/letsencrypt/live/mail.<domain>.com/` have all required certificate. (Mount location through **docker-compose.yml**, by adding 
```
volumes:
   - /etc/letsencrypt:/etc/letsencrypt
```
- Now their is a good point to catch here. For some reason, my SMTP port for SSL/TLS was not exposed to docker container. So I added that missing port **465** in **docker-compose.yml** like this:
```
 ports:
     - "25:25"
     - "143:143"
     - "587:587"
     - "993:993"
     - 465:465
```

## Part that is still not working
When tried to connect through Gmail android client. I got `Invalid certificate error`.  So I clicked on proceed to proceed further without a certificate. Now all the emails are TLS encrypted but this error while setting-up ports still persist.

**Comment by smitpatelx** [comment]
# Manual Fix

### /etc/postfix/main.cf

> smtpd_tls_cert_file=/etc/ssl/certs/ssl-cert-snakeoil.pem //lets encrypt fullchain.pem
> smtpd_tls_key_file=/etc/ssl/private/ssl-cert-snakeoil.key // privkey.pem

### /etc/dovecot/conf.d/10-ssl.conf

> ssl_cert = </etc/dovecot/ssl/dovecot.pem //lets encrypt fullchain.pem
> ssl_key = </etc/dovecot/ssl/dovecot.key // privkey.pem

I think something is wrong, so I check certificate and key used by dovecot and postfix and I found these certificates don't match letsencrypt certificates. Which means dovecot and postfix are not using a letsencrypt certificate.

So I manually fixed it to see if it's working.

```
// Fix it by docker exec <container name or id>

docker exec <container name or id> cp /etc/letsencrypt/live/<domain_name>/fullchain.pem /etc/ssl/certs/ssl-cert-snakeoil.pem
docker exec <container name or id> cp /etc/letsencrypt/live/<domain_name>/privkey.pem /etc/ssl/private/ssl-cert-snakeoil.key
docker exec <container name or id> cp /etc/letsencrypt/live/<domain_name>/fullchain.pem /etc/dovecot/ssl/dovecot.pem
docker exec <container name or id> cp /etc/letsencrypt/live/<domain_name>/privkey.pem /etc/dovecot/ssl/dovecot.key
docker exec <container name or id> service postfix reload
docker exec <container name or id> dovecot postfix reload
docker exec <container name or id> service postfix status
docker exec <container name or id> dovecot postfix status

```
This did the trick. Now I don't face any errors.

> **I will post a script for `DockerFile` in this issue.  That will fix all the issues I guess.**

**Comment by fbartels** [comment]
Thanks for your investigation and the steps you have taken to get it running on your side.

My suspicion is still that when you are using a wildcard certificate the part at https://github.com/tomav/docker-mailserver/blob/master/target/start-mailserver.sh#L979-L983 does not match the wildcard entry (since its not the $HOSTNAME, but just a part of it). Your steps seem to indicate this as well.

**Comment by smitpatelx** [comment]
So you are saying that my hostname must be `mail.<domain>,com`? Can we add `$HOSTNAME + $DOMAINNAME` or something? What's the right thing to do?  @fbartels

**Comment by smitpatelx** [comment]
Fixed it. My letsenctypt spelling was wrong. Silly I know 😂😂🤣. 

Have one question can my mail server name be just `<domain>.com`? I tried to override in **env-mailserver**, but then SMTP was not working. It was giving errors. @fbartels

**Comment by fbartels** [comment]
> you are saying that my hostname must be mail.<domain>.com

no, not `mail.`, but it needs to run on a subdomain in general.

> Can we add $HOSTNAME.$DOMAINNAME or something?

that is in general already there. There is a pr to make the letsencrypt part more dynamic at https://github.com/tomav/docker-mailserver/pull/1301, but it has become stale.

> It was giving errors.

its always easier to give precise error description, because these can then be looked up in the code from which additional insights can be gained.

**Comment by smitpatelx** [comment]
I want an email server on `example.com` instead of current 'mail.netdevv.com'. Is there any way to override this? Or is no option or configuration available for that? @fbartels

**Comment by erik-wramner** [comment]
I don't know if it has changed, but in the late 90s using only the domain without host name as an MX record was problematic. I guess you want to use a single certificate for the domain without resorting to wild cards?

If it is important for you, make it work first. Then when you know what you want to change we can see if it can make it into the build. Otherwise the typical configuration is to use a host for the MX record and that same host for the TLS certificate.

**Comment by smitpatelx** [comment]
Ok, I don't think its a bad practice to user subdomains but most email clients by default add `domain.com` by default. So every time I log in to any mail agent like Gmail for android, I don't want to write **mail.domain.com**. So I am questioning this approach of running mail servers on subdomains.

I know some people might think to use a subdomain if they don't want to show their IP address for main domain. But I think if we are using the same VM for all domains and subdomains, it doesn't make any sense to use a subdomain. 

**What is the real reason for using a subdomain? Am I wrong or missing something important?**

**Comment by smitpatelx** [comment]
# Suggestion for future release: 
Add `letsencrypt_wildcard=true` to **env-mailserver** so if it's true look for only `domain.com` in **/etc/letsencrypt/live**. By default you can set it false. Can anyone who might be expert in shell scripting. Please add a loop to check for this new variable **letsencrypt_wildcard**

**Comment by erik-wramner** [comment]
Well, I'm not an expert on this, but I'd use the term host rather than sub-domain. The MX record should point to a host and normally you have host.domain.root (for example mail.example.com). It is possible to define an A record that treats just the domain as a host without a host name, but that is not the classical way of doing things. If you know that you have and intend to have a single host in your domain you can skip the host name, but you may run into problems with programs that don't support it and expect to get a non-blank host name.

In short the real reason for using a host (not sub-domain) is that the Internet standards were written that way by the ancients. Things have changed, but some of us are conservative. Using a host name is the safe choice.

**Comment by aendeavor** [comment]
A solution to this specific issue has been found, I guess. I can also confirm that Traefik's ACME Wildcard Certificates work fine.
